src/gsts3sink.c

```diff
@@ -734,15 +787,35 @@ gst_s3_sink_do_flush (GstS3Sink * sink)
 
     // Set current position to end of data in current buffer
     const gsize buffer_start = sink->current_pos - sink->buffer_pos;
+    if (sink->buffer_from_cache) // potentially bad assumption about to happen
+      sink->buffer_size = sink->config.buffer_size;
     sink->buffer_pos = sink->buffer_size;
     sink->current_pos = buffer_start + sink->buffer_pos;
 
-    if (sink->buffer_size != sink->config.buffer_size &&
-        sink->current_pos < sink->upload_size) {
+    if (sink->uploader_needs_complete) {
+      GST_DEBUG_OBJECT(sink, "uploader needs completion; doing that then copy-upload to buffer start");
+      gst_s3_uploader_complete(sink->uploader);
+      gst_s3_destroy_uploader(sink);
+      sink->uploader = GST_S3_SINK_GET_CLASS((gpointer*) sink)->uploader_new (&sink->config);
+      sink->uploader_needs_complete = FALSE;
 
+      // copy-upload from 0->buffer_start - 1
+      gst_s3_uploader_upload_part_copy(
+        sink->uploader,
+        sink->config.bucket,
+        sink->config.key,
+        0,
+        buffer_start - 1
+      );
+    }
+
+    if (!sink->buffer_from_cache &&
+        sink->buffer_size != sink->config.buffer_size &&
+        sink->current_pos < sink->upload_size)
+    {
       const gsize bytes_to_read = MIN(sink->config.buffer_size - sink->buffer_size, sink->upload_size - sink->current_pos);
       GST_TRACE_OBJECT(sink, "Post-filling %ld bytes, range: %ld-%ld", bytes_to_read, sink->current_pos,
-      sink->current_pos + bytes_to_read);
+        sink->current_pos + bytes_to_read);
```

Comment:
Fixing formatting.  A lot of the whitespace adds/removes in this file were because they didn't have an indent where one would have been nice, like this multi-line call.